tools/check/ut.go

```diff
@@ -0,0 +1,529 @@
+		return nil, withTrace(err)
+	}
+	for _, x := range xxx {
+		tasks = append(tasks, task{pkg, x, false})
+	}
+
+	yyy, err := listOldTestCases(pkg)
+	if err != nil {
+		fmt.Println("list old test case error", pkg, err)
+		return nil, withTrace(err)
+	}
+	for _, y := range yyy {
+		tasks = append(tasks, task{pkg, y, true})
+	}
+	return tasks, nil
+}
+
+func listPackages() ([]string, error) {
+	cmd := exec.Command("go", "list", "./...")
+	ss, err := cmdToLines(cmd)
+	if err != nil {
+		return nil, withTrace(err)
+	}
+
+	ret := ss[:0]
+	for _, s := range ss {
+		if !strings.HasPrefix(s, modulePath) {
+			continue
+		}
+		pkg := s[len(modulePath)+1:]
+		if skipDIR(pkg) {
+			continue
+		}
+		ret = append(ret, pkg)
+	}
+	return ret, nil
+}
+
+type numa struct {
+	cpu     string
+	numactl bool
+}
+
+func (n *numa) worker(wg *sync.WaitGroup, ch chan task) {
+	defer wg.Done()
+	for t := range ch {
+		start := time.Now()
+		if err := n.runTestCase(t.pkg, t.test, t.old); err != nil {
+			fmt.Println("run test case error", t.pkg, t.test, t.old, time.Since(start), err)
+		}
+	}
+}
+
+func (n *numa) runTestCase(pkg string, fn string, old bool) error {
+	_, file := path.Split(pkg)
+	exe := path.Join(workDir, pkg, file+".test")
+	var cmd *exec.Cmd
+	if n.numactl {
+		cmd = n.testCommandWithNumaCtl(exe, fn, old)
+	} else {
+		cmd = n.testCommand(exe, fn, old)
+	}
+	cmd.Dir = path.Join(workDir, pkg)
+	_, err := cmd.CombinedOutput()
+	if err != nil {
+		// fmt.Println("run test case error", pkg, fn, string(output))
+		return err
+	}
+	return nil
+}
+
+func (n *numa) testCommandWithNumaCtl(exe string, fn string, old bool) *exec.Cmd {
+	if old {
+		// numactl --physcpubind 3 -- session.test -test.run '^TestT$' -check.f testTxnStateSerialSuite.TestTxnInfoWithPSProtoco
+		return exec.Command(
+			"numactl", "--physcpubind", n.cpu, "--",
+			exe,
+			"-test.timeout", "20s",
+			"-test.cpu", "1", "-test.run", "^TestT$", "-check.f", fn)
+	}
+
+	// numactl --physcpubind 3 -- session.test -test.run TestClusteredPrefixColum
+	return exec.Command(
+		"numactl", "--physcpubind", n.cpu, "--",
+		exe,
+		"-test.timeout", "20s",
+		"-test.cpu", "1", "-test.run", fn)
+}
+
+func (n *numa) testCommand(exe string, fn string, old bool) *exec.Cmd {
+	if old {
+		// session.test -test.run '^TestT$' -check.f testTxnStateSerialSuite.TestTxnInfoWithPSProtoco
+		return exec.Command(
+			exe,
+			"-test.timeout", "20s",
+			"-test.cpu", "1", "-test.run", "^TestT$", "-check.f", fn)
+	}
+
+	// session.test -test.run TestClusteredPrefixColum
+	return exec.Command(
+		exe,
+		"-test.timeout", "20s",
+		"-test.cpu", "1", "-test.run", fn)
+}
+
+func skipDIR(pkg string) bool {
+	skipDir := []string{"br", "cmd", "dumpling"}
+	for _, ignore := range skipDir {
+		if strings.HasPrefix(pkg, ignore) {
+			return true
+		}
+	}
+	return false
+}
+
+func buildTestBinary(pkg string) error {
+	// go test -c
+	cmd := exec.Command("go", "test", "-c", "-vet", "off")
```

Comment:
We now have a plan to introduce `gotestsum`: https://github.com/pingcap/tidb/pull/29848

It should be considered together @tangenta

Reply:
The `ut` tool introduced in this PR will not replace the old CI pipeline immediately, there many works to do before make it ready, including the coverage.

Reply:
Got it.